Inline text formatting

barklund opened the pull request:
This is an initial implementation of inline text formatting.

Fixes #413.
Fixes #1007.
Fixes #925.

### **TL;DR**:
It works for all the properties: bold, italic, underline font weight, letter spacing and text color, but there's still some stuff to do - however everything can and will be fixed! Also, there's a couple of open questions, that need be answered to fully implement.

### Goal
The goal is to be able to update font weight (and bold as a subset of that), italic, underline, text color and letter-spacing both inline and on the entire text field.

This also includes that the text panel is aware of which of these features are currently set to which values while editing the text field - so when selecting text, that "B" button is correctly pressed or not based on whether the currently selected text is all (some) bold.

All of this has to work as before for multiple text fields as well and things should work when pasted too.

### Implementation details

#### Structure
This introduces a new provider located just inside the `CanvasProvider` in `Workspace`.

Also, the entire rich text editor is moved to a new component that is then utilized by the text edit component. We might not need this component elsewhere, but the split made sense.

One caveat is, that the rich text provider directly accesses the `editingElementState` from the `CanvasProvider`. It would probably be cleaner to pass this in all the way from the text edit component. I might do that in the future.

The text edit component now **only** updates the HTML on unmount.

#### Formatting
This PR also completely overrules draftjs' built-in formatting and creates a completely new formatting setup with custom style names. There's a parser, a serializer and a set of rules for display.

This also re-implements the logic as to what it means to press "italic" (both by shortcut and button) by answering the question _"is current selection italic or not?"_ explicitly. It seems like a trivial question, and kind of is for italic, but for bold is a lot more complex - what if some characters are semibold, others are black - what happens when you press the "bold" button? All this is answered in code in `styleManipulation`.

These same rules will be applied to the entire text field (in upcoming posts) as well as partial selection (which works now for a subset of the properties).

### Status

This is the progress so far:
- [x] Bold by shortcut and button
- [x] Italic by shortcut and button
- [x] Underline by shortcut and button
- [x] Bold on entire text field
- [x] Italic on entire text field
- [x] Underline on entire text field
- [x] Font weight on selection
- [x] Font weight on entire text field
- [x] All the above "entire text field" rules applied to multiple selected text fields.
- [x] Deprecate old properties not used anymore (`fontStyle`, `textDecoration`, `bold`, `fontWeight`).
- [x] All the above for the "empty" selection. That is, if you're in edit mode but haven't selected anything, just placed your cursor somewhere, and press "bold" and "red", the next text you write at that point should now correctly start out as "bold" and "red". Seems easy enough to do though - just check if selection is collapsed and if so, apply [inline style change to override](https://draftjs.org/docs/api-reference-editor-state#setinlinestyleoverride) rather than selection.
- [x] Letter spacing on selection
- [x] Letter spacing on entire text field (+multi-selection)
- [x] Color on selection
- [x] Color on entire text field (+multi-selection)
- [x] Deprecate old properties not used anymore (`color`, `letterSpacing`).
- [x] Display `(multiple)` in font weight dropdown when it's the case
- [x] Migrate old properties to new inline syntax (which will include a fun new exercise on merging diverging settings in `fontWeight` versus `bold` - I'm thinking that `if fontWeight=400 and bold` wrap everything in `font-weight:700`, else use actual font weight). Plus intensive unit test!
- [x] Refactor and cleanup code so text formatting options are in verticals, not horizontals (adding color as a new option required changing 7-8 files, should just be a single file).
- [x] Display "faux-selection" background when editor is unfocused
- [x] Unit test formatters
- [x] Unit test style manipulation functions
- [x] Make sure color presets still work for text fields – both for entire text field and selection. And both for adding and (re)using presets.
- [x] Update and amend unit tests for color presets
- [x] Fix font weights in text panel presets
- [ ] Export proper font summary for text field for proper client-side font inclusion as added in #1273
- [ ] Unit test text style panel section
- [ ] Unit test font weight panel section
- [ ] Unit test text color panel section
- [ ] Integration test all this in connection

Known bugs:

- [x] ~Proper parsing of multiple styles applied to the same character (due to https://github.com/sstur/draft-js-utils/pull/155 going unmerged) - will fix with `patch-package` in a future commit) [bug can be seen by making some text both italic and bold, exiting edit mode and re-entering edit mode]~ Fixed in 854510f7
- [ ] Pasting formatted content doesn't correctly "reformat" to our new styles (will fix in future commit)
- [ ] When creating inline style override using controls in the design panel, previous inline style overrides will be lost. Example: Move cursor to some non-bold, non-italic word without selecting anything. Press "bold" button (sets bold in inline override). Now press "italic" button (removes bold from inline override due to focus loss, but then sets italic override). This can be fixed, but it's a bit tricky - it's about remember the inline style override when losing focus, and merging old style back in when focus is regained. There's an open PR for this in https://github.com/facebook/draft-js/pull/2022, but we can work around it without waiting for it to get merged.
- [ ] When a single text element is selected, typing a single letter on the keyboard should make the text element go into edit mode with focus and the typed key replacing the entire old content. This is currently broken for unknown reasons.
- [ ] When inline-selection covers text of all the same color but in a text field with multiple colors, it is not currently possible to add the currently selected color to neither text style nor text color presets.

Possible optimizations:

- [ ] Output proper semantic HTML (currently using `<span />` with class names and inline styles for all formatting) – proper semantic DOM is not easy due to limitations to `draft-js-utils`, where only one node can be returned per styled section, where something both bold and italic should have two surrounding nodes (thus the single `<span />` used now with multiple styles and class names).
- [ ] Don't directly access canvas provider from inside rich text provider.

Open questions:

- ~What should the font weight drop down display, when multiple font weights are present in selection? (my current idea is to push a disabled `<option>(multiple)</option>` to the front of the list and mark it as selected)~ added as a proper task above
- ~What should the letter spacing input display, when multiple letter spacings are present in selection? (currently going for `(multiple)` as used in inputs elsewhere)~ already handled by component
- ~What should the font color component display, when multiple color are present in selection? (will implement some kind of `(multiple)` logic which currently doesn't exist for color display as far as I know).~ already handled by component
- What happens to invalid font weights when you change font family? Not sure how to handle this yet.
- Should you be able to reset color and letter spacing by keyboard shortcut?
- Should you be able to set different font weights (other than bold or non-bold) by keyboard shortcut?

### Demo
![inline-text-editing-v1](https://user-images.githubusercontent.com/637548/80144452-80b4e880-857c-11ea-9275-a5ced84c4357.gif)
(when you see the text change without the mouse being over the buttons, it's because I'm using shortcuts - forgot to include the on-screen key-logger for this video)

Commit:
WIP: First partially working version of inline text formatting

Commit:
Merge branch 'master' into feature/inline-rich-text

barklund commented on assets/src/edit-story/components/richText/provider.js:
```diff
@@ -0,0 +1,196 @@
+    const workingState = EditorState.forceSelection(oldState, selection);
+    const newState = updater(workingState);
+    setEditorState(newState);
+    setForceFocus(true);
```
This is part of the magic happening. When you press a button (or use a dropdown, input field or color picker) outside the editor, the editor loses focus, but can remember the selection. Here we simply reapply that selection, then perform the formatting as requested, and then ask the editor to refocus the input field - moving focus away from whatever the user pressed elsewhere.

This however is not going to work with color picker, as the text color should probably update as the user clicks around in the color picker, but **not** move focus away from the color picker, as that will close it.

Similarly for the letter spacing input, as you can type and see it take effect in real time, without losing focus in the letter spacing input field.

barklund replied:
And this came up now with letter spacing - as soon as you type a single letter, focus moves back to the editor. Will have to exclude force focus for this one.

barklund replied:
Fixed in 4a28da7d.

barklund commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  return [...styles];
+}
```
This one is a mix of other functions doing the same thing - iterating over all selected characters across whichever blocks and styles they're in. See [this](https://github.com/webdeveloperpr/draft-js-custom-styles/blob/f3e6b533905de8eee6da54f9727b5e5803d53fc4/src/index.js#L8-L52) and [this](https://github.com/facebook/draft-js/issues/602#issuecomment-584676405) for reference.

Commit:
Added fix for importing elements with multiple styles

barklund commented on patches/draft-js-import-element+1.4.0.patch:
```diff
@@ -0,0 +1,26 @@
+               break;
+             }
+ 
+diff --git a/node_modules/draft-js-import-element/lib/stateFromElement.js b/node_modules/draft-js-import-element/lib/stateFromElement.js
```
This file doesn't actually need fixing on my setup, but don't know if that's just a local thing. Fixed both exported JS files to be safe. Ping @swissspidy.

Commit:
Added the option to toggle b/i/u for entire text field

Commit:
Cleaned up the structure a bit

Commit:
Added support for font weight and multi-select

Commit:
Update tests now that some values are not in use

Commit:
Adding `npx` to postinstall to see if CI approves

swissspidy commented on package.json:
```diff
@@ -31,7 +31,9 @@
     "mousetrap": "^1.6.5",
+    "patch-package": "^6.2.2",
     "polished": "^3.5.2",
+    "postinstall-postinstall": "^2.1.0",
```
This is only needed for yarn?

barklund replied:
Oh great, I'll remove it then.

barklund replied:
Fixed in b0ba3011

Commit:
Removed `postinstall-postinstall`

Commit:
If selection is collapsed when applying style change, apply to inline override

Commit:
Reorganised selection manip and added letter spacing

Commit:
Disable force focus when setting letter spacing

Actually the whole `forceFocus` variable was superfluous - `EditorState.forceSelection` autimatically forces focus to the editor. And selection was remembered, just not visible while unfocused, so things still work as expected.

Commit:
Added inline text color formatting

Commit:
Fix multiple reducer for patterns, and fix tests

Commit:
Fixed font and font weight dropdowns

Font weight dropdown simply gets a new placeholder, as the placeholder will only be shown, when there's multiple font weights selected.

For now, font weight logic is removed from font selector (but must be readded in a new version later), and tests are updated.

Commit:
Remove deprecated types from text element

Commit:
Some extra cleanup of deprecated properties

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/font.js:
```diff
@@ -101,11 +90,10 @@ function FontControls({ selectedElements, pushUpdate }) {
             <DropDown
               data-testid="font.weight"
               ariaLabel={__('Font weight', 'web-stories')}
+              placeholder={__('(multiple)', 'web-stories')}
```
I assume this should placeholder should only be done when `selectedElements.length > 1`?

barklund replied:
Nope.

Any text selection, even in a single text field, can have more than one font weight. And will only show the placeholder specifically in case of "multiple". "Multiple" here does not mean "multiple text fields with different values", but just "multiple font weights present in selection", which can be anything from a selection inside a text field, to an entire text field to multiple text fields.

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js:
```diff
@@ -0,0 +1,138 @@
+import { MULTIPLE_VALUE } from '../../form';
+
+function isEqual(a, b) {
+  const isPattern = typeof a === 'object' && (a.type || a.color);
```
`isEqualColor`? Or if this is used for other objects, would we ever run into cases with `typeof null === 'object'` and `typeof [] == 'object'`?

barklund replied:
Well, here it's used specifically for a known set of properties, and there's also the requirement of a property to be present on the object, so arrays won't be possible nor trigger this property.

But yes, `null` might be an issue. Will fix that one.

The naming of the function is a "generic" equality function, but only for primitives **or** colors - not other types of objects.

dvoytenko replied:
This statement would go very nicely in the jsdoc :)

barklund replied:
Fixed and commented in 41e69548.

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js:
```diff
@@ -0,0 +1,138 @@
+  const aStyle = generatePatternStyles(a);
+  const bStyle = generatePatternStyles(b);
+  const keys = Object.keys(aStyle);
+  return keys.every((key) => aStyle[key] === bStyle[key]);
```
Nit, but might be also easier to do `convertToCSS() === convertToCSS()`.

barklund replied:
Fixed in 41e69548.

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js:
```diff
@@ -0,0 +1,138 @@
+} from '../../richText/htmlManipulation';
+import { MULTIPLE_VALUE } from '../../form';
+
+function isEqual(a, b) {
```
jsdoc.

barklund replied:
Added in 41e69548.

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js:
```diff
@@ -0,0 +1,138 @@
+  return keys.every((key) => aStyle[key] === bStyle[key]);
+}
+
+function reduceWithMultiple(reduced, info) {
```
jsdoc

barklund replied:
Added in 41e69548.

dvoytenko commented on assets/src/edit-story/components/panels/textStyle/useRichTextFormatting.js:
```diff
@@ -0,0 +1,138 @@
+    (updater, ...args) =>
+      pushUpdate(
+        ({ content }) => ({ content: updater(content, ...args) }),
+        true
```
Not all of the field updates are auto-submit, right? E.g. `letter-spacing` is not auto-submit. It's just a normal numeric input.

barklund replied:
I now updates the text field as soon as a value is entered. But the new content is (as always) not submitted until the edit mode is exited.

dvoytenko replied:
Can you explain how this will go for letter-spacing? The sequence of input:

1. `150` (initial value, indicating `150%`)
2. `15` (the right-most character is deleted)
3. `1` (the right-most character is deleted)
4. `''` (the right-most character is deleted, this is not an empty string)

barklund replied:
The input handling in the input component here is not optimal - it'll push its changes on every change, and then re-read the current value from the real input, resulting in empty value not being allowed, so it's un-intuitive to type some values (e.g. you cannot clear the text field). That's an optimization of the input component though, not the text formatter IMHO.

dvoytenko replied:
Ok. Let's proceed and we'll need to see how we can follow up on this. For now this is one of few items that allows intermediate values. It might be best to do this update via the form's `usePresubmitHandler`, but can be done as a follow up.

dvoytenko commented on assets/src/edit-story/components/richText/customConstants.js:
```diff
@@ -0,0 +1,65 @@
+ * Letter spacing uses PREFIX-123 for the number 123
+ * and PREFIX-N123 for the number -123.
+ */
+export const letterSpacingToStyle = (ls) =>
```
This is totally cool, but just a note: both weight and letter-spacing are essentially `numericToStyle(name, value)` and `styleToNumeric(name, style)`. The fact that one can be negative is just a detail.

barklund replied:
Yeah, I'll rework these a bit and they'll probably reuse the same functionality when that happens.

barklund replied:
Refactored and reorganized into [`numericToStyle`/`styleToNumeric`](https://github.com/google/web-stories-wp/pull/1323/files#diff-01ff94c89f89c22b31e8f53c5cd74328R22-R39) used for [weight](https://github.com/google/web-stories-wp/pull/1323/files#diff-c062d7168620373d6b39280d2038e7b8R32-R38) and [letter spacing](https://github.com/google/web-stories-wp/pull/1323/files#diff-36cedac683c0b4b3afc0a28a582d4804R28-R34).

dvoytenko commented on assets/src/edit-story/components/richText/customConstants.js:
```diff
@@ -0,0 +1,65 @@
+export const SMALLEST_BOLD = 600;
+export const DEFAULT_BOLD = 700;
+
+export const isStyle = (style, prefix) =>
```
Just out of curiosity, would it have worked if we simply used here CSS notation. E.g. `BOLD-700` becomes a custom style `font-weight:700` (or `fontWeight:700`) instead? And so on. If possible, this could possibly simplify some of the downstream code.

barklund replied:
Good idea. I'll think about it as I'll be refactoring this part of the codebase.

barklund replied:
I have not changed this part as of yet. It's still an open option though.

dvoytenko commented on assets/src/edit-story/components/richText/customExport.js:
```diff
@@ -0,0 +1,108 @@
+      // Italic
+      if (style === ITALIC) {
+        return {
+          classes: [...classes, 'italic'],
```
Are classes mainly for output? Are they strictly necessary? I'm kind of really warming up to the idea of just using inline styles for this always. Especially since many styles are non-enumerable.

barklund replied:
They're only for re-parsing the content back in to the proper style names. It's easier to pattern match on the classes rather than the styling, was my thinking.

dvoytenko replied:
Ok. If it's noticeable easier - sure. But `style` is pretty easy to interrogate as well, no? E.g. both `classList` and `style` can be cycled through easily, and both can also do membership checks just as easily.

barklund replied:
I've removed classes in refactor - using only styles now. See e.g weight [import](https://github.com/google/web-stories-wp/pull/1323/files#diff-c062d7168620373d6b39280d2038e7b8R40-R49) and [export/display](https://github.com/google/web-stories-wp/pull/1323/files#diff-c062d7168620373d6b39280d2038e7b8R51-R61).

dvoytenko commented on assets/src/edit-story/components/richText/customImport.js:
```diff
@@ -0,0 +1,98 @@
+
+            case 'letterspacing': {
+              const ls = element.style.letterSpacing;
+              const lsNumber = ls.split(/[a-z%]/)[0] || 0;
```
Doesn't `parseInt()` do it already?

barklund replied:
Relying on `parseInt` to trim trailing non-numerical characters seems dirty to me. But yes it does.

dvoytenko replied:
Ok. Let's keep then.

barklund replied:
Using [implicit trim now with a comment](https://github.com/google/web-stories-wp/pull/1323/files#diff-36cedac683c0b4b3afc0a28a582d4804R38-R39).

dvoytenko commented on assets/src/edit-story/components/richText/customImport.js:
```diff
@@ -0,0 +1,98 @@
+              return null;
+          }
+        })
+        .filter((style) => Boolean(style));
```
Are 0's never ok?

barklund replied:
`0`'s? In what way?

dvoytenko replied:
E.g. for `letter-spacing`, a `Boolean(0) => false`?

barklund replied:
The style would in that case be `LETTERSPACING-0` which is still _truthy_. This check is just to root out some rogue class names, that don't correspond to any styles.

But other than that, the style `LETTERSPACING-0` should never be present. Just like `WEIGHT-400` as both are the default styling.

dvoytenko commented on assets/src/edit-story/components/richText/util.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+let contentBuffer = null;
+export const draftMarkupToContent = (content, bold) => {
```
We don't need `bold` here anymore, right?

barklund replied:
This function will die (or be rewritten completely). It's still legacy.

dvoytenko replied:
Yep. Commented on something like this later as well.

barklund replied:
This has been converted to a generic util in 8685d11d.

dvoytenko commented on assets/src/edit-story/components/richText/customInlineDisplay.js:
```diff
@@ -0,0 +1,63 @@
+function customInlineDisplay(styles) {
+  return styles.toArray().reduce((css, style) => {
+    // Italic
+    if (style === ITALIC) {
```
These all seem word-for-word `customExport.js` minus classes?

barklund replied:
Yes, they'll be merged in refactor.

barklund replied:
Completely merged now - [`customExport`](https://github.com/google/web-stories-wp/pull/1323/files#diff-8628a016619e622cf15fd6bf32d09cc7R27-R41) and [`customInlineDisplay`](https://github.com/google/web-stories-wp/pull/1323/files#diff-23f240017de355129a38853b10d045c2R22-R27) are almost identical now except for minutiae.

dvoytenko commented on assets/src/edit-story/components/richText/editor.js:
```diff
@@ -0,0 +1,106 @@
+    }
+  }, [onChange, getContentFromState, editorState]);
+
+  // Set focus when initially rendered.
```
IMHO, this should be done in the caller via forwarded ref.

barklund replied:
You think so? I can change that, no problem. I don't feel strongly either way.

dvoytenko replied:
Hmm. I know we only have one of these. But seems like it's not abnormal to potentially have a couple in DOM shown and it'd be odd for them to fight over the focus. `TextEdit` however knows it's always, by flow, should be focused when mounted.

barklund replied:
Fixed in 8097c3f.

dvoytenko commented on assets/src/edit-story/components/richText/editor.js:
```diff
@@ -0,0 +1,106 @@
+  // Push updates to parent when state changes
+  useEffect(() => {
+    const newContent = getContentFromState(editorState);
+    if (newContent) {
```
Is a empty string a possible valid value here?

barklund replied:
Hm, good question. Can you clear a text field and if so, what should happen? Currently it'll revert to what it was before entering edit mode because of this exact line.

barklund replied:
I think this is okay for now. We can look at it later if need be.

dvoytenko replied:
When text is cleared, we should just display placeholder again. Ok to defer.

dvoytenko commented on assets/src/edit-story/components/richText/editor.js:
```diff
@@ -0,0 +1,106 @@
+  useEffect(() => {
+    const newContent = getContentFromState(editorState);
+    if (newContent) {
+      onChange(newContent);
```
I have a bit of a resistance doing onChange via side effect when a clean user event is available somewhere here. I can be convinced otherwise, though, if there's a good reason. Could you pls explain?

barklund replied:
editorState can change based on many different factors (typing; keyboard shortcut for bold, italic, underline, undo, redo; and clicking/typing in design panel). This seemed to me to be the nicest way to capture them all.

dvoytenko replied:
Yeah. Let's keep going in this direction.

dvoytenko commented on assets/src/edit-story/components/richText/editor.js:
```diff
@@ -0,0 +1,106 @@
+  // Allow parent to focus editor and access main node
+  useImperativeHandle(ref, () => ({
+    focus: () => editorRef.current?.focus?.(),
+    getNode: () => editorRef.current?.editorContainer,
```
Nit: you can also do `get node() {return ...}` if your semantics are more of a getter.

barklund replied:
I didn't really know what the expectation was for this kind of API - I couldn't find many examples of `useImperativeHandle` in the wild, so there doesn't seem to be a best-practice here. I can go either way.

dvoytenko replied:
Just a FYI. Either way is fine by me as well. Generally `useImperativeHandle` either returns an actual element, or exposes APIs that are kind of similar to DOM elements in nature. E.g. a FancyVideo component can return the actual video element so that a caller can call `play()` and `pause()`. But YoutubeVideo component doesn't have a video element (since it's an iframe) and would often simulate video element's API.

barklund replied:
I kept this as-is.

dvoytenko commented on assets/src/edit-story/components/richText/htmlManipulation.js:
```diff
@@ -0,0 +1,72 @@
+import customExport from './customExport';
+import { getSelectionForAll } from './util';
+
+function getSelectAllStateFromHTML(html) {
```
Minimal jsdocs would really be helpful here.

barklund replied:
Added in 97a91cf3.

dvoytenko commented on assets/src/edit-story/components/richText/htmlManipulation.js:
```diff
@@ -0,0 +1,72 @@
+export function toggleBoldInHTML(html, flag) {
+  return updateAndReturnHTML(html, toggleBold, flag);
+}
+export function setFontWeightInHTML(html, weight) {
```
Nit: +1 line.

barklund replied:
Restructured completely. These two functions don't exist explicitly anymore but handled by [generalized HTML formatters](https://github.com/google/web-stories-wp/pull/1323/files#diff-4bb929db50ae52c99fcd00b92fa60ec7R47-R60).

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  DEFAULT_BOLD,
+} from './customConstants';
+
+function getPrefixStyleForCharacter(styles, prefix) {
```
Ditto: minimal jsdoc.

barklund replied:
Done in 0081a79c.

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  const endKey = selection.getEndKey();
+  const endOffset = selection.getEndOffset();
+  let hasMoreRounds = true;
+  while (hasMoreRounds) {
```
This looks like a `do {...} while(key !== endKey)` cycle?

barklund replied:
It is. But I find `do-while`'s to cause more confusion than good - the next dev might not understand it at first glance and break it. And it doesn't save many cycles here.

dvoytenko replied:
Ok. Leaving this up to you. No strong opinion here.

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  return allIsBold;
+}
+
+export function toggleBold(editorState, flag) {
```
Just to confirm: there's no such API call as `toggleBold(editorState, 800)`?

barklund replied:
`bold` is a boolean. `fontWeight` is a number.

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+    typeof flag === 'boolean'
+      ? weightToStyle(flag ? DEFAULT_BOLD : NORMAL_WEIGHT)
+      : weightToStyle(
+          Math.max.apply(null, [DEFAULT_BOLD].concat(getWeights(styles)))
```
`Math.max(...[DEFAULT_BOLD].concat(...))`

barklund replied:
Fixed [at some point](https://github.com/google/web-stories-wp/pull/1323/files#diff-c062d7168620373d6b39280d2038e7b8R93).

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  const getBoldToSet = (styles) =>
+    typeof flag === 'boolean'
+      ? weightToStyle(flag ? DEFAULT_BOLD : NORMAL_WEIGHT)
+      : weightToStyle(
```
A slight asymmetry here is confusing. For the first one we could have either `700` or `400`. But for the else-expression it can only be `>=700`. But then we also have `shouldSetBold`, right? If we already have `shouldSetBold`, shouldn't we even have an `if` here?

barklund replied:
There's an `if` in `togglePrefixStyle`. It'll only use `getBoldToSet` if `shouldSetBold` returns true.

dvoytenko replied:
Ok. So we don't need this `if` here? It's simply `return weightsToStyle(DEFAULT.concat(...))`, right?

barklund replied:
Ah, now I understand the question.

No, this is actually needed for a single special case. If you select two textfields, both not completely bold but one of them partially black (900), and we don't have this condition, they'd end up being different bolds after pressing the button - the one without any bold goes to 700 and the other to its own max weight = 900.

And that felt like a mistake, so this ensures that if multiple text fields are pressed, bold toggle can only ever go to one of two weights: 400 or 700 - even if one of the text fields contain a "bolder bold" inside it.

barklund replied:
Let me know if this is not obvious. I have `https://github.com/google/web-stories-wp/pull/1323/files#diff-c062d7168620373d6b39280d2038e7b8R77-R101` (was done during testing to simplify tests).

dvoytenko replied:
GOt it.

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  return togglePrefixStyle(editorState, WEIGHT, shouldSetBold, getBoldToSet);
+}
+
+export function getFontWeight(editorState) {
```
Are all of these methods always work on selection? If not, it'd be useful to have that reflected in the name. If that's the case, though - no need.

barklund replied:
Yes, everything in this file works on the current selection as represented by the editor state selection object. But this will get reworked too - to be split in verticals rather than horizontals (so there'e be one file for all the logic for `italic`, another for all `fontWeight`/`bold`, etc).

dvoytenko replied:
Ok. And currently the full-text updates are `editorState.forceSelection(selectAll)`, right?

barklund replied:
Yes, actually the full-text updates are roughly `content = export(toggleBold(selectAll(import(content))));` as there is no actual editor to invoke these changes on - [this happens in `htmlManipulation`](https://github.com/google/web-stories-wp/pull/1323/files/47df628514fefc5c55a10424e0c0e1585edd6520#diff-4bb929db50ae52c99fcd00b92fa60ec7R38-R49).

dvoytenko commented on assets/src/edit-story/components/richText/styleManipulation.js:
```diff
@@ -0,0 +1,319 @@
+  const styles = getPrefixStylesInSelection(editorState, WEIGHT);
+  const weights = getWeights(styles);
+  if (weights.length > 1) {
+    return MULTIPLE_VALUE;
```
Is `MULTIPLE_VALUE` useful on this API level? Or would it be better to have an array returned actually?

barklund replied:
Yeah, I was debating that myself. I didn't actually need the list of values anywhere though, so I decided on the simpler return. There might be a case for returning the array, but I can't see it.

dvoytenko replied:
Ok. Let's move with this. I think it's clear enough so that if array is needed we can refactor.

dvoytenko commented on assets/src/edit-story/components/richText/useSelectionManipulation.js:
```diff
@@ -0,0 +1,103 @@
+  const setLetterSpacingInSelection = useCallback(
+    (ls) =>
+      updateWhileUnfocused(
+        (s) => setLetterSpacing(s, ls),
```
Please spell out `s` and/or `ls`.

barklund replied:
This specific line doesn't exist anymore - have spelled out all args elsewhere.

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -78,7 +73,7 @@ const Span = styled.span`
 
 const BackgroundSpan = styled(Span)`
-  color: transparent;
+  color: transparent !important;
```
Why?

barklund replied:
The background span must only include the background color - no foreground colors.

Before just overriding color with `transparent` was sufficient, as color was set on the entire text field.

But now that color can be set on individual elements inside the text field, this `transparent` most override those too.

dvoytenko replied:
Not sure I understand. Could you point me to some example to understand this part better?

barklund replied:
Which part? Why we need to have a background layer with transparent text color or why we have multiple elements with their own text color inline property?

dvoytenko replied:
The "background layer with transparent color" part.

dvoytenko replied:
ping on this

barklund replied:
This should illustrate *why* we need two layers of text - we need background-only on the bottom layer and foreground-only on the top layer (as this is not an uncommon design to strive for - the multi-layer one that is):
https://codepen.io/barklund/pen/WNQdrmM

dvoytenko replied:
Ok. I got this part. But can't we use the same utils we have for style manipulation to remove the inline color? E.g. `const transparentContent = setColor(context, 'transparent')`?

barklund replied:
Hm, we could actually. Not directly like that as it expects a pattern object, but we could give it a `{color: {r: 0, g: 0, b: 0:, a: 0 } }`.

Or even better implement [my note here about making black the default color](https://github.com/google/web-stories-wp/pull/1323/files#diff-7c25b43d56bb77d2a8190efa2f7e0a2cR80-R81) and then use black to indicate, that no color style should be applied anywhere (thus default'ing to the class-based transparent value).

barklund replied:
I ended up doing the latter in b63842a and then changed to using the existing functionality in b6468b0 - which is a much better solution. Thanks! ✋ 5️⃣ 🎉

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -147,7 +132,7 @@ function TextDisplay({
               {...props}
               dangerouslySetInnerHTML={{
-                __html: draftMarkupToContent(content, bold),
+                __html: draftMarkupToContent(content),
```
That method looks rather misnamed. It's really just `cleanContent()` or such.

barklund replied:
Yes, it's about to be completely removed (or reduced). I just hadn't gotten around to it yet.

barklund replied:
Added to todo as "cleanup".

dvoytenko replied:
Resolving this since we have another thread open on this exact item.

dvoytenko commented on assets/src/edit-story/elements/text/edit.js:
```diff
@@ -243,23 +163,28 @@ function TextEdit({
 
   // Update content for element on focus out.
-  useFocusOut(textBoxRef, updateContent, [updateContent]);
+  //useFocusOut(textBoxRef, updateContent, [updateContent]);
```
Remove

barklund replied:
Fixed in 8097c3fb.

dvoytenko commented on assets/src/edit-story/elements/text/output.js:
```diff
@@ -130,7 +118,7 @@ export function TextOutputWithUnits({
   const backgroundTextStyle = {
     ...textStyle,
-    color: 'transparent',
+    color: 'transparent !important',
```
We definitely can't do `!important` here.

barklund replied:
Why not?

dvoytenko replied:
I believe AMP doesn't allow `!important` styles.

barklund replied:
Oh, that's fair. Then we need to manually find and replace inline color styling though. But it's not too hard.

dvoytenko replied:
Ping on this.

barklund replied:
So, I've fixed this by removing inline color styling. It's not a super pretty solution though, so any alternative is very welcome! I tried going the regexp route first, but it ran into too many issues (as HTML is not regular as you know).

Please see c5adfecf

barklund replied:
Changed it in b63842a and b6468b0 - which is a much better solution.

dvoytenko reviewed (changes requested):
This looks very very good!

miina commented on assets/src/edit-story/elements/text/index.js:
```diff
@@ -36,15 +36,10 @@ export const defaultAttributes = {
   fontSize: 36,
-  fontStyle: 'normal',
   backgroundColor: createSolid(196, 196, 196),
-  color: createSolid(0, 0, 0),
```
Don't forget to rework the color presets logic in case the color attribute gets removed with this PR.

barklund replied:
Oh yes, thanks. Yeah, that work has to be done too! And you should be able to apply color presets for both entire text field or a selection, I assume. I'll make sure it works for preset panel too.

barklund replied:
Added to todo.

barklund replied:
Fixed in 03aa64b and 1b4f9de

Commit:
Added new migration with test (also added `jest-extended`)

Commit:
Merge branch 'master' into feature/inline-rich-text

miina commented on assets/src/edit-story/elements/text/index.js:
```diff
@@ -36,15 +36,10 @@ export const defaultAttributes = {
   bold: false,
   fontFamily: 'Roboto',
   fontFallback: ['Helvetica Neue', 'Helvetica', 'sans-serif'],
-  fontWeight: 400,
```
Ah, one more note: currently `Story_Post_Type(.php)` is [using `fontWeight`](https://github.com/google/web-stories-wp/blob/fdd822ab3e0dd9927b8a12b9bbc623d4c7d951a7/includes/Story_Post_Type.php#L452-L455) for loading the correct font in the FE. This will be moved to client-side with #1273 (cc @swissspidy) but something to consider anyway for font generation.

barklund replied:
@miina, is this fixed or does it need follow-up?

miina replied:
This needs a follow-up.
For testing:
1. Create a Text block
2. Assign Mukta Mahee font, Extra Light, toggle Italic
3. Save and preview
4. See the font loaded in the FE: https://fonts.googleapis.com/css2?display=swap&family=Mukta+Mahee which includes only `normal` font-style, it should, however, put together correct font link to fetch the correct font variants: italic + extra light in this case (variant `[1, 200]`).

We actually have an issue which should include a fix for that as well: https://github.com/google/web-stories-wp/issues/923

The issue was created thinking of the previous draftjs bold/italic option only, however, now it should fix it for the full text element -- it's the same logic needed there anyway -- partial selection or full.

That issue is already in the Sprint backlog, I might take a look at it meanwhile based on your PR.

dvoytenko replied:
I believe this is being addressed by @obetomuniz in https://github.com/google/web-stories-wp/pull/1275

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Refactored formatters to verticals rather than horizontals - much cleaner result

Commit:
Fixed reducer and added comments

Commit:
Added jsdocs to html manipulation

Commit:
Removed classes from migration and added test for nested tags

Commit:
Fix tests for text style now that export does not have classes anymore

Commit:
Moved focus management to edit component

Commit:
Create general `getValidHTML` util

dvoytenko commented on assets/src/edit-story/components/richText/editor.js:
```diff
@@ -0,0 +1,93 @@
+  useUnmount(clearState);
+
+  // Allow parent to focus editor and access main node
+  useImperativeHandle(ref, () => ({
```
Let's use explicit `[]`, for `deps` here.

barklund replied:
Done in 3a809040.

dvoytenko commented on assets/src/edit-story/components/richText/formatters/letterSpacing.js:
```diff
@@ -0,0 +1,99 @@
+function elementToStyle(element) {
+  const isSpan = element.tagName.toLowerCase() === 'span';
+  // This will implicitly strip any trailing unit from the value - it's assumed to be em
+  const letterSpacing = parseFloat(element.style.letterSpacing, 10);
```
What's `10` here? Leftover from `parseInt`?

barklund replied:
🤦 (fixed in e2d2f236)

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -147,7 +132,7 @@ function TextDisplay({
               {...props}
               dangerouslySetInnerHTML={{
-                __html: draftMarkupToContent(content, bold),
+                __html: getValidHTML(content),
```
Case could be made if we always correctly save `content`, then it will always be valid in the data model and no need to validate it each time. The validation is a bit expensive given the HTML construction/parsing/reformatting/etc.

barklund replied:
@miina fixed issues elsewhere with malformed HTML, right? Not sure how it happened, but until we're sure it doesn't happen again, I'd rather keep this in (but yes, it's a bit expensive - we might want to memoize it).

dvoytenko replied:
Yes, I believe all of the issues with malformed HTML have been resolved. And we certainly should have `getValidHTML()` called whenever we call `updateProperties({content})` - there can't be that many of them. So, yeah, we can check that and drop `getValidHTML()` and reading sites.

dvoytenko commented on assets/src/edit-story/elements/text/output.js:
```diff
@@ -146,7 +134,7 @@ export function TextOutputWithUnits({
               style={backgroundTextStyle}
               dangerouslySetInnerHTML={{
-                __html: draftMarkupToContent(content, bold),
+                __html: getValidHTML(content),
```
Ditto: if we are convinced that we save content correctly everywhere - no reason to validate it again here.

barklund replied:
Closing this one in favor of the other discussion.

dvoytenko commented on assets/src/edit-story/migration/migrations/v0015_inlineTextProperties.js:
```diff
@@ -0,0 +1,150 @@
+    return wrapped;
+  }
+
+  const justBold = 'font-weight: 700';
```
Maybe I'm misreading this, but it looks like `700` gets handled by the `if (globalWeight) {}` above? In either case, I assume `isBold === false` is a valid case here too?

barklund replied:
There's a difference between the two handlers:

- The global weight in the top wraps the entire text field in a single `<span style="font-weight: X;">` while removing any and all inline `<strong>` tags. Hence `stripTag` and `wrapWithSpan`.
- The inline replacement in the bottom replaces existing `<strong>` tags with a similar inline `<span style="font-weight: X;">`, but not globally - just inline and in multiple places. Hence `replaceTagWithSpan`.

So yes, if `isBold` is not `true` and if `fontWeight` is missing or `400`, we'll just convert any existing `strong` tags with _bolded_ `span`'s - and if there aren't any of those tags, nothing happens.

dvoytenko replied:
Got it.

dvoytenko reviewed (changes requested):
This is getting very close. Should we move it out of draft?

Commit:
Added deps to `useImperativeHandle`

Commit:
Fixed `parseFloat` args

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Added faux-selection style and fixed various bugs

Commit:
Fixed typo

Commit:
For now, strip formatting when pasting

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Remove weird npm scripts

Commit:
Using a less nice matcher

Commit:
Added tests for formatters

miina reviewed (dismissed):
Will add some found bugs as comments.
(And as you know, the color presets / fonts aren't working currently, mentioning it just to ensure the PR doesn't get merged accidentally without these!)

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Set default font color to black - all other colors are fixed inline

Commit:
Added unit tests for style manipulation

These are slightly integrated with draft-js' editor state, as it was too complex too mock and this is actually a better test IMHO.

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Fixed color input (allow multiple value string)

Commit:
Fixed color picker to default to solid black (rather than transparent black)

Commit:
Fixed null value for input

Commit:
Fix color preset to work for new ITF formatting

Commit:
Added full support for text color and text style presets

Commit:
Inlined font weights for text panel presets

miina commented on assets/src/edit-story/components/panels/stylePreset/test/panel.js:
```diff
@@ -217,7 +214,7 @@ describe('Panels/StylePreset', () => {
             textColors: [],
             fillColors: [],
-            textStyles: [STYLE_PRESET],
+            textStyles: [{ color: TEST_COLOR_2, ...STYLE_PRESET }],
```
Just a color alone would not be part of `textStyles`, it would be part of `textColors`.

barklund replied:
But it's not a color alone - it's color and whatever's in `STYLE_PRESET` (assuming it's non-empty)?

miina replied:
Whoops, saw incorrectly as if these were two separate array items.

miina commented on assets/src/edit-story/components/panels/stylePreset/test/panel.js:
```diff
@@ -204,7 +201,7 @@ describe('Panels/StylePreset', () => {
       getTextPresets.mockImplementation(() => {
         return {
-          textStyles: [STYLE_PRESET],
+          textStyles: [{ color: TEST_COLOR_2, ...STYLE_PRESET }],
```
Same here as below:
Just a color alone would not be part of textStyles, it would be part of textColors.

miina commented on assets/src/edit-story/components/panels/stylePreset/test/panel.js:
```diff
@@ -155,9 +153,7 @@ describe('Panels/StylePreset', () => {
-            color: [TEST_COLOR_2],
-            backgroundTextMode: BACKGROUND_TEXT_MODE.NONE,
-            font: TEXT_ELEMENT_DEFAULT_FONT,
+            content: '<span style="color: rgba(2, 2, 2)">Content</span>',
```
Q: Why remove the other props, the `font` and `backgroundTextMode`? That's basically for ensuring that even though those exist, the logic should still consider only the color for the preset -- default font and `BACKGROUND_TEXT_MODE.NONE` should be ignored.

barklund replied:
Re-added in 16dbf27f.

Commit:
Fix focus out to check click target in capture phase

dvoytenko commented on assets/src/edit-story/components/library/panes/text/textPane.js:
```diff
@@ -42,9 +42,11 @@ const PRESETS = [
     id: 'heading',
     title: __('Heading', 'web-stories'),
-    content: __('Heading', 'web-stories'),
+    content: `<span style="font-weight: 700">${__(
```
Can we do a utility that does a very quick and limited (but side-effect-free) construction of content for our constants? Then if/when we switch the format slightly - that'd be easier to keep up-to-date.

barklund replied:
Not sure what you mean? If we move` __()` to another function, it'll not be completely side-effect free, as we have tooling depending on this syntax.

dvoytenko replied:
Hmm. Never mind. I'm just slightly concerns about the text constant here (or translation) yielding no-valid HTML (e.g. someone would insert a "&" or "<"). Maybe we should juts wrap `__()` inside the `escapeHTML()`?

dvoytenko commented on assets/src/edit-story/components/richText/draftUtils.js:
```diff
@@ -0,0 +1,52 @@
+      offsetIndex < offsetEnd;
+      offsetIndex++
+    ) {
+      styleSets.push(characterList.get(offsetIndex).getStyle());
```
I'd assume we'd want to do a `Set` here? Or just check for dups?

barklund replied:
Is an array (allowing duplicates) or sets of styles (each not allowing duplicates). It wouldn't matter if the outer array was a set IMHO.

I'll add jsdoc describing this in e295b37f.

dvoytenko commented on assets/src/edit-story/components/richText/provider.js:
```diff
@@ -0,0 +1,138 @@
+}
+
+RichTextProvider.propTypes = {
+  children: PropTypes.oneOfType([
```
`children: PropTypes.node`

barklund replied:
Fixed in 870f2bdf.

dvoytenko commented on assets/src/edit-story/components/panels/stylePreset/panel.js:
```diff
@@ -119,24 +121,46 @@ function StylePresetPanel() {
     ]
   );
 
+  const extraPropsToAdd = useRef(null);
```
Really not a big fun of this. But can't suggest anything better at this point. IMHO, we just need to merge with 65 files and counting and fix forward.

barklund replied:
Merge what? Not sure what you're referring to here? It's pretty ugly having to do this, but I don't see the consequences that you mention?

Commit:
Merge branch 'master' into feature/inline-rich-text

# Conflicts:
#	assets/src/edit-story/migration/migrate.js
#	package.json

Commit:
Re-add default props to test

Commit:
Added jsdoc for draftUtils

Commit:
Fix children proptype

Commit:
Add better error handlign to faux selection removal

Commit:
Added jsdocs for complex style manipulation functions

dvoytenko reviewed (changes requested):
I posted a few more responses on the existing threads. But otherwise my only question is what to do with `!important` styles for the output (and lesser extend the editor as well). Otherwise, I'm ready to approve this.

Commit:
Added function to strip relevant inline styling

Commit:
Merge branch 'master' into feature/inline-rich-text

# Conflicts:
#	package.json

dvoytenko commented on assets/src/edit-story/elements/text/output.js:
```diff
@@ -139,14 +129,18 @@ export function TextOutputWithUnits({
   if (backgroundTextMode === BACKGROUND_TEXT_MODE.HIGHLIGHT) {
+    const foregroundContent = getValidHTML(content);
+    const backgroundContent = getValidHTML(content, (node) =>
+      removeInlineStyle(node, 'color')
```
I'm a bit lost why we need a separate ~PR~ util. Also commented on the same in https://github.com/google/web-stories-wp/pull/1323#discussion_r419765427. We already have DraftJS utils to do this. But not a blocker to merge.

dvoytenko replied:
/s/PR/util

barklund replied:
Replied above.

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -142,14 +128,18 @@ function TextDisplay({
   });
 
   if (backgroundTextMode === BACKGROUND_TEXT_MODE.HIGHLIGHT) {
+    const foregroundContent = getValidHTML(content);
```
Hmm. Looking at this again. I think this really needs useMemo.

barklund replied:
Implemented a reusable solution with memoization in b6468b0d.

dvoytenko reviewed (approved):
Approved, but at least one thing I'd like to change to add useMemo on valid content.

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Added new function to compare patterns and used where applicable

Commit:
Added ignore black to set color style as it is default

Commit:
Changed highlight to use existing style manipulation

Reverted changes to `getValidHTML` and reused the existing HTML manipulation functions.

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -142,14 +128,21 @@ function TextDisplay({
+    const foregroundContent = content;
+    // Setting the text color of the entire block to black essentially removes all inline
+    // color styling allowing us to apply transparent to all of them.
+    const backgroundContent = useMemo(
```
We can't put a hook inside the `if () {}`.

barklund replied:
Yeah, my bad. Fixed in 13b6fdac.

dvoytenko commented on assets/src/edit-story/elements/text/display.js:
```diff
@@ -142,14 +128,21 @@ function TextDisplay({
+    // Setting the text color of the entire block to black essentially removes all inline
+    // color styling allowing us to apply transparent to all of them.
+    const backgroundContent = useMemo(
+      () => getHTMLFormatters().setColor(content, createSolid(0, 0, 0)),
```
Is the `createSolid` missing an alpha 0?

barklund replied:
No - we apply solid opaque black, which is default meaning that [_no color style_ will be applied to the entire text field](https://github.com/google/web-stories-wp/pull/1323/files#diff-7c25b43d56bb77d2a8190efa2f7e0a2cR81-R83). Thus allowing us to overwrite the whole thing with `color:transparent`.

dvoytenko reviewed (approved):
Two more things noticed.

Commit:
Merge branch 'master' into feature/inline-rich-text

Commit:
Removed conditional hook
